src/game/client/draw/w3dsupplydraw.cpp

```diff
@@ -0,0 +1,103 @@
+/**
+ * @file
+ *
+ * @author Jonathan Wilson
+ *
+ * @brief W3D Supply Draw Module
+ *
+ * @copyright Thyme is free software: you can redistribute it and/or
+ *            modify it under the terms of the GNU General Public License
+ *            as published by the Free Software Foundation, either version
+ *            2 of the License, or (at your option) any later version.
+ *            A full copy of the GNU General Public License can be found in
+ *            LICENSE
+ */
+#include "w3dsupplydraw.h"
+#include "drawable.h"
+
+void W3DSupplyDrawModuleData::Build_Field_Parse(MultiIniFieldParse &p)
+{
+    // clang-format off
+    static const FieldParse dataFieldParse[] = {
+        { "SupplyBonePrefix", &INI::Parse_AsciiString, nullptr, offsetof(W3DSupplyDrawModuleData, m_supplyBonePrefix) },
+        { nullptr, nullptr, nullptr, 0 },
+    };
+    // clang-format on
+
+    W3DModelDrawModuleData::Build_Field_Parse(p);
+    p.Add(dataFieldParse, 0);
+}
+
+NameKeyType W3DSupplyDraw::Get_Module_Name_Key() const
+{
+    static const NameKeyType nk = g_theNameKeyGenerator->Name_To_Key("W3DSupplyDraw");
+    return nk;
+}
+
+void W3DSupplyDraw::Update_Draw_Module_Supply_Status(int max, int current)
+{
+    W3DModelDraw::Update_Draw_Module_Supply_Status(max, current);
+    Utf8String prefix = Get_W3D_Supply_Draw_Module_Data()->m_supplyBonePrefix;
+
+    if (m_totalSupplyBoneCount == -1) {
+        m_totalSupplyBoneCount = Get_Drawable()->Get_Pristine_Bone_Positions(prefix, 1, nullptr, nullptr, 0x7FFFFFFF);
+        m_currentSupplyBoneCount = m_totalSupplyBoneCount;
+    }
+
+    int newcount = GameMath::Ceil((float)m_totalSupplyBoneCount * ((float)current / (float)max));
+    newcount = std::min(newcount, m_totalSupplyBoneCount);
+
+    if (newcount != m_currentSupplyBoneCount) {
+        int minbone = std::min(m_currentSupplyBoneCount, newcount);
+        int maxbone = std::max(m_currentSupplyBoneCount, newcount);
+        bool hide = newcount < m_currentSupplyBoneCount;
+        int bonecount = minbone + 1;
+        std::vector<ModelConditionInfo::HideShowSubObjInfo> subobjs;
+
+        while (bonecount <= maxbone) {
+            char name[16];
+            sprintf(name, "%s%02d", prefix.Str(), bonecount);
```

Comment:
sprintf_s makes sense but bone names are fixed by the w3d format at 16 characters therefore using a 16 character array is correct here.

Reply:
But we append digits, at least 2, so it can be longer.

Reply:
```
{ "SupplyBonePrefix", &INI::Parse_AsciiString, nullptr, offsetof(W3DSupplyDrawModuleData, m_supplyBonePrefix) },
```

From my understanding, `m_supplyBonePrefix` can be any length at this point. There is nothing that caps the input.

Reply:
If someone puts a prefix in the ini that's too long, things will break regardless of the size of the array because it would produce bone names that are longer than the valid length of a bone in a w3d file so it's impossible for those to be valid bones. We should not be setting subobj.sub_obj_name to a name that is longer than the valid length of a bone in a w3d file.

Reply:
Ok that is fair. Then we should add an assert in the field parse for `m_supplyBonePrefix <= 13` (16 - null terminator - 2 digits) to make it obvious that things break then.